in_place_upgrade_guest: automates the case that verifies in place upgrade on a guest

rh-jugraham opened the pull request:
Case ID: VIRT-293757

Depends on: https://github.com/avocado-framework/avocado-vt/pull/4070

Automates the case that verifies that an in place upgrade can be successfully completed on a guest. This case has already been automated by /tp-qemu/qemu/tests/in_place_upgrade.py - this specific test in tp-libvirt exists such that an in place guest upgrade test can be accomplished without depending on 'kar'.

This test specifically supports upgrading from rhel X-1 -> rhel X on a rhel X host.

Tests passing:
```
(.libvirt-ci-venv-ci-runtest-ZbmaDy) [root@ampere-mtsnow-altra-09 ~]# avocado run --vt-type libvirt --vt-omit-data-loss --vt-machine-type arm64-mmio in_place_upgrade_guest --vt-connect-uri qemu:///system
JOB ID     : 3947dcc69ad9e030e06a3828ae35e8470aa508dc
JOB LOG    : /var/log/avocado/job-results/job-2025-02-24T15.09-3947dcc/job.log
 (1/1) type_specific.io-github-autotest-libvirt.in_place_upgrade_guest.rhel9_to_rhel10.0: STARTED
 (1/1) type_specific.io-github-autotest-libvirt.in_place_upgrade_guest.rhel9_to_rhel10.0: PASS (1233.92 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/log/avocado/job-results/job-2025-02-24T15.09-3947dcc/results.html
JOB TIME   : 1235.61 s
```

rh-jugraham commented:
@dzhengfy @Yingshun Could you help review this PR? 

A few notes:
1. I added a 64k kernel workaround for now that can be removed when/if there is a change
2. I was considering adding a new test type as this is not technically a libvirt specific test; however, I wasn't sure if it made sense to create a new test type for 1 test
3. I wasn't sure how to best add in version specific information - I added it in the cfg under the rhel9_to_rhel10.0 variant (similar to the tp-qemu ipu test) but unsure if there is an alternative

iccaszhulili commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,16 @@
+        - rhel9_to_rhel10.0:
+            pre_major_release = "9"
+            target_major_release = "10"
+            target_release = "RHEL-10.0"
```
Hi, Julia

As I heard from dzheng, that we just need to test upgrade on RHEL10 host, so , we I think we may just need to check the host major and minor version, then we can get the target_release. And I think what we need to input is the pre_minor_release, since the major release is always RHEL.(X-1), we do not support upgrade from RHEL.(X-2) to RHEL.X in one shot.

rh-jugraham replied:
@iccaszhulili 

I'm not sure if I am understanding your comment correctly, so let me know if I misunderstood. What I realized is that we know the release before upgrade (RHEL-X-1.y) and we know the target major release (RHEL-X) but we don't know the target minor release. So I tried modifying the code such that everything is calculated (no need to hard code what doesn't need to be) except for the target minor release, which needs to be given with, for example, the variant `to_RHELX.0` for target minor release RHEL-X.0 (such as migrating rhel-9.6 --> rhel10.0).

Updated tests passing:
```
(.libvirt-ci-venv-ci-runtest-ZbmaDy) [root@ampere-mtsnow-altra-09 ~]# avocado run --vt-type libvirt --vt-omit-data-loss --vt-machine-type arm64-mmio in_place_upgrade_guest --vt-connect-uri qemu:///system
JOB ID     : d1ac967df701dbdc57a74e246ade58a97ac0a4aa
JOB LOG    : /var/log/avocado/job-results/job-2025-02-25T16.45-d1ac967/job.log
 (1/1) type_specific.io-github-autotest-libvirt.in_place_upgrade_guest.to_rhelX.0: STARTED
 (1/1) type_specific.io-github-autotest-libvirt.in_place_upgrade_guest.to_rhelX.0: PASS (994.40 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/log/avocado/job-results/job-2025-02-25T16.45-d1ac967/results.html
JOB TIME   : 996.08 s
```

iccaszhulili replied:
Hi, Julia 

> @iccaszhulili
> 
> I'm not sure if I am understanding your comment correctly, so let me know if I misunderstood. What I realized is that we know the release before upgrade (RHEL-X-1.y) and we know the target major release (RHEL-X) but we don't know the target minor release. 
I think we can get the minor release from the host, for example, in RHEL10.1 release testing cycle, we need to test upgrade from RHEL9.7 to RHEL10.1. During testing, we will provision a host with RHEL10.1 OS, then we can know that we need upgrade the guest to RHEL10.1, the guest OS version will be aligned with the host OS version.

rh-jugraham replied:
This makes more sense, thanks!

iccaszhulili commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,16 @@
+    type = in_place_upgrade_guest
+    release_check_cmd = "cat /etc/redhat-release"
+    kernel_check_cmd = "uname -r"
+    pagesize_check_cmd = "getconf PAGESIZE"
```
Hi, Julia 

Do we need to do the following steps
```
# yum -y install kernel-64k 
# k=$(echo /boot/vmlinuz*64k) 
# grubby --set-default=$k --update-kernel=$k --args="crashkernel=2G-:640M" 
# reboot
```
before doing upgrade. Is there an existing 64k image for testing?

iccaszhulili commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,14 @@
+    list_kernel_cmd = "grubby --info=ALL | grep ^kernel"
+    default_kernel_cmd = "grub2-set-default"
+    variants:
+        - to_rhelX.0:
```
Hi, Julia

I am Okay with it. Just to mention that, we may need to do some settings in CI. We need to ask CI to provision a guest with latest RHEL(X-1) OS. We need to make sure CI knows the minor release of RHEL(X-1).

dzhengfy replied:
We have various guest in different versions in US/PEK2 storage server. So I think we just fetch it as we want.

rh-jugraham replied:
@dzhengfy This does make more sense to me (we have a rhelX.z machine and we are testing upgrading from rhelX-1.y --> X.z). However, how do we make sure the CI knows the minor release of rhelX-1.y - how do we know which guest is the latest to fetch? And what is the best way to fetch it?

rh-jugraham replied:
Update: understood - the test will assume that it already has a guest at version X-1.y on a host with version X.z.

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,134 @@
+        if vm.is_alive():
+            vm.destroy(gracefully=False)
+        test.log.info("Restoring vm %s...", vm.name)
+        vmxml_backup.sync()
```
If I did not miss anything, I did not see any changes to vm xml. So we can remove this line.

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,134 @@
+
+    # 1) Prepare a running guest
+    vm = env.get_vm(vm_name)
+    vmxml_backup = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
```
If I did not miss anything, I did not see any changes to vm xml. So we can remove this line.

dzhengfy reviewed (commented):
Others LGTM

rh-jugraham commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,14 @@
+    release_check_cmd = "cat /etc/redhat-release"
+    kernel_check_cmd = "uname -r"
+    pagesize_check_cmd = "getconf PAGESIZE"
+    compose_url = "http://download.eng.bos.redhat.com"
```
@dzhengfy Done - created a PR in ci-shared-datas. One question: is it an issue that the test has `baseurl={compose_url}/rhel-{target_major_release}/nightly...`? I'm not sure if the full path like this belongs in the code but I had struggled defining it completely in the cfg (especially getting vm_arch_name).

smitterl commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,11 @@
+- in_place_upgrade_guest:
+    type = in_place_upgrade_guest
+    release_check_cmd = "sed 's/[^0-9.]*//g' /etc/redhat-release"
```
It's generally preferrable to handle strings in Python code to make debugging easier. It might be unlikely in this specific case but the pattern or the place for the searched-for information might change, in that case the release check would return an empty string but the logs wouldn't record what the output was.
So, I'd prefer to see you read `/etc/redhat-release` as string and use python module `re`.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,138 @@
+    if step == "upgrade":
+        leapp_cmd = params.get("leapp_upgrade_cmd")
+
+    leapp_status, leapp_output = session.cmd_status_output(leapp_cmd, timeout=900)
```
This line is longer than the usual 80 characters. The context (function name, command variable) already indicate it's related to leapp, so `status, output` should suffice.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,138 @@
+    if leapp_status == 0:
+        test.log.info("Leapp %s executed successfully" % step)
+    else:
+        test.log.debug("Leapp %s output: %s" % (step, leapp_output))
```
The leapp output with --debug can be really long.
Alternatively, it could make more sense to copy the leapp pre/upgrade log into the test folder, origin `/var/log/leapp/leapp-(pre)upgrade.log` target.

The test might handle it similar to the `LibvirtdDebugLog` in https://github.com/avocado-framework/avocado-vt/blob/master/virttest/env_process.py#L1012
But even if not, the debug log is also stored in the test log directory so it could at least help understand how the code can construct the path to the test log directory where to copy the file to.

rh-jugraham replied:
I do know that the vm logs do include the the full output of the (pre)upgrade cmd so technically nothing needs to be done. The reason why I had originally decided to add it into specifically the debug log was because I know that, for example, when running tests with jenkins, you often only get access to the debug log. However, given that this test will likely not be run in that context, having this output in the debug log is unnecessary as it is very long.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,138 @@
+            enabled=0
+            gpgcheck=0
+        """)
+        session.cmd("cat <<EOF > %s %s\nEOF" % (upgrade_repos_path, upgrade_repos_content))
```
As mentioned above, it's usually better to use the python library instead of bash in this case:

with(upgrade_repos_path...) as f:
     f.write(f"""
...
gpgcheck=0
""")
```

rh-jugraham replied:
@smitterl How would I do so on the vm since `with open() as f` tries to open the file on the host - I haven't been able to find an alternative for the guest.

rh-jugraham replied:
Updated! Now using `with open() as f` to create the file on the host then copying it over to the guest.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,138 @@
+            vm.start()
+        session = vm.wait_for_login()
+
+        # 2) Calculate target release and pagesize
```
We often try to apply clean code principles. One demands putting logic into smaller units (functions) and maintain the same abstraction level.

The test code could hence be written as:

```
from_release, to_release = get_release_info(release_check_cmd)
check_support(from_release, to_release)
utils_package.package_install(...)
configure_leapp_tool(compose_url, to_release)
pre_ugprade_info.append(get_guest_info(session, guest_cmd)) # << this can be generalized for other conditions
if pre_upgrade_succeeds():
    run_upgrade_and_reboot()
check_upgrade_post_conditions(pre_upgrade_info, vm, params)

...
def check_upgrade_post_conditions(pre_upgrade_info, vm, params):
...
    session = vm.wait_for_login()
    check_kernel_version(session)
    post_upgrade_info.append(guest_info(...))
    confirm_are_equal(pre_upgrade_info, post_upgrade_info)

```

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,138 @@
+        session.close()
+
+    finally:
+        # Destroy VM
```
Is this comment necessary?

smitterl reviewed (changes requested):
Thanks for this. I think this can be improved to be more generally useful, esp. on more architectures.

rh-jugraham commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,11 @@
+    leapp_preupgrade_cmd = "leapp preupgrade --debug --no-rhsm"
+    leapp_upgrade_cmd = "leapp upgrade --debug --no-rhsm"
+    list_kernel_cmd = "grubby --info=ALL | grep ^kernel"
+    default_kernel_cmd = "grub2-set-default"
```
Good call - I created a [PR](https://github.com/avocado-framework/avocado-vt/pull/4070) for that!

rh-jugraham commented:
> Thanks for this. I think this can be improved to be more generally useful, esp. on more architectures.

@smitterl Thank you for the detailed feedback! I believe I have addressed all comments or added a question if confused.

iccaszhulili commented on libvirt/tests/cfg/in_place_upgrade_guest.cfg:
```diff
@@ -0,0 +1,10 @@
+    type = in_place_upgrade_guest
+    release_check_cmd = "cat /etc/redhat-release"
+    kernel_check_cmd = "uname -r"
+    pagesize_check_cmd = "getconf PAGESIZE"
```
@yicui Is this command better than `uname -r`?

iccaszhulili replied:
@dzhengfy `getconf PAGESIZ` is better than `uname -r`?

dzhengfy replied:
@iccaszhulili I think both are ok.` uname -r ` is mainly for checking the kernel version, not pagesize here.

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+def get_release_info(params, session):
+    """
+    Get the pre release and target release versions
+    """
```
Please add docstring like  `:return: tuple, (source  version, target  version)` as well as the parameters

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+def extract_release_version(release_output):
+    """
+    Extract the release version from the full release output
+    """
```
Add dostring for paremeter

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+def check_support(test, pre_release, target_release):
+    """
+    Check whether the test supports upgrading from pre_release to target_release
+    """
```
Complete the docstring, thanks

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+    release_check_cmd = params.get("release_check_cmd")
+    pre_release = extract_release_version(session.cmd_output(release_check_cmd))
+    target_release = extract_release_version(process.run(release_check_cmd, shell=True).stdout_text)
+    return (pre_release, target_release)
```
Could you add more debug infomation for your returned values?

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+        f.write(upgrade_repos_content)
+
+    vm.copy_files_to(host_path=tmp_upgrade_repos_path, guest_path=upgrade_repos_path)
+
```
Could you add more debug infomation to describe what this function does?

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+def run_leapp_cmd(test, params, session, step):
+    """
+    Run the leapp (pre)upgrade cmd
+    """
```
Complete docstring, please

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    Verify whether the upgrade succeeded by checking
+    the release version, the kernel version, and the page size
+    """
```
Complete docstring, please

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+    if target_release not in post_release_output:
+        test.fail("The guest after upgrade should be at rhel %s, but is at %s" %
+                  (target_release, post_release_output))
+
```
I suggest we add `else` clause to display debug information like what result is as expected.

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+    post_kernel_output = session.cmd_output(kernel_check_cmd)
+    if (".el%s" % target_major_release) not in post_kernel_output:
+        test.fail("The guest after upgrade should be at kernel version .el%s, but is at %s" %
+                  (target_major_release, post_kernel_output))
```
ditto

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+
+    post_page_size = session.cmd_output(pagesize_check_cmd).strip()
+    if target_page_size != post_page_size:
+        kernel_version = get_kernel_version(test, params, session, target_release, target_page_size)
```
Please add some INFO like this is a workaround so far and should be removed after leapp tool support 64k on aarch64. Thanks

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+        if target_page_size != post_page_size:
+            test.fail("Page size before upgrade (%s) is not the same after upgrade (%s)" %
+                      (target_page_size, post_page_size))
+
```
`Else` clause

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+def get_kernel_version(test, params, session, target_release, target_page_size):
+    """
+    Get the kernel version given the release version and page size
+    """
```
Docstring, please

dzhengfy commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+    try:
+        kernel = [kernel for kernel in kernels if release in kernel and page_size in kernel][0]
+        kernel_version = kernel.split("/")[-1].strip('\"')
+        return kernel_version
```
Before return, please add more debug info.

rh-jugraham commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,183 @@
+        [APPSTREAM]
+        name=APPSTREAM
+        baseurl={compose_url}/rhel-{target_major_release}/nightly/RHEL-{target_major_release}/latest-RHEL-{target_release}/compose/AppStream/{vm_arch_name}/os/
+        enabled=0
```
@dzhengfy Yes, I believe so? But could you clarify this question?

rh-jugraham replied:
I see, makes sense!

dzhengfy reviewed (commented):
I am fine with other codes. Just waiting for the clarification about enable repo or not

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+            test.log.info("Guest page size after upgrade and workaround is %s, as expected" % post_page_size)
+
+
+def get_kernel_version(test, params, session, target_release, target_page_size):
```
IMO, this function should live in https://github.com/avocado-framework/avocado-vt/pull/4070 or where ever the `update_vm_default_kernel` function lives because it can help others who use that function. They will likely have to re-implement the logic presented here, or similar logic, to achieve that.

rh-jugraham replied:
Could do so. 

Do you know how I could get more information about kernel naming conventions/differences between architectures? Since I am used to aarch64, and this code currently only runs for aarch64, I don't know how many assumptions I am making that would not translate well onto other architectures (besides of course page size being an arm specific thing).

smitterl replied:
I don't know the kernel naming conventions but I'm not sure if it's absolutely necessary. By using the proposed function from 

your code could have a switch:

```
if params.get("use_64k") == "yes":
    kernel_pattern += ".*64k"

kernel_version = utils_test.get_available_kernel_path(kernel_pattern)
update_vm_default_kernel(vm, kenrel_version, reboot=True, guest_arch_name=machine.platform())
```

rh-jugraham replied:
I understand now. I was assuming that the helper function would get the kernel path given the os release version, but I realized that I can simply get the kernel path(s) given any regex pattern

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+                      (target_page_size, post_page_size))
+        test.log.info("Attempting temporary workaround to update the guest default kernel")
+        kernel_version = get_kernel_version(test, params, session, target_release, target_page_size)
+        utils_test.update_vm_default_kernel(vm, kernel_version, timeout=900, guest_arch_name=vm_arch_name)
```
If you trigger the reboot here through the default parameter, IMO you should still set it explicitly 'reboot=True' because the next time this code is reviewed the reviewer might not be familiar with the `update_vm_default_kernel` function and its defaults. However, here it's crucial.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+
+    finally:
+        if vm.is_alive():
+            vm.destroy(gracefully=False)
```
Usually, we expect tests to restore the VM as it was before. This is not possible here and is only achieved by avocado-vt's `backup_image` and `restore_image` configuration values (s. https://github.com/avocado-framework/avocado-vt/blob/master/virttest/shared/cfg/base.cfg#L582)

You can decide to back up and restore the original image file in your test case or at least add a WARNING at the start of the `run` docstring. If the backup/restore functionality is not configured (AFAICT it isn't s. `base.cfg` above) then the test can't be rerun, the test data has to be set up again.

You could also check for the configuration and raise a "TestError" to inform the person who wants to execute this test to please consider these configuration parameters. Unfortunately, the functionality is implemented in `virttest/env_process.py` so you can't force the image to be backed up.

rh-jugraham replied:
So I am not sure if this is what you are referring to, but when running this test, there is the following line in the debug log after the test has completed:

`avocado.virttest.storage storage          L1189 DEBUG| Copying /var/lib/avocado/data/avocado-vt/images/jeos-27-aarch64.qcow2.backup -> /var/lib/avocado/data/avocado-vt/images/jeos-27-aarch64.qcow2`

There are also the following options already set by default, as I can see in the first few lines of the debug.log, which I am assuming are responsible for the above code:
```
backup_image_before_testing = yes
restore_image_after_testing = yes
```

So I believe that the VM is being restored (though not obvious from my implementation as this relies on defaults)? I can rerun this test multiple times and everything seems fine.

smitterl replied:
Yes, this is what I'm referring to bug apparently there's also some confusion on both sides because there seem to be several parameters with similar names. Sorry in advance for the long reply.

`backup_image`
-----------------------

`backup_image` is set to `no` per default in `base.cfg` our CI doesn't modify this.
On review, I noticed this parameter is apparently not used everywhere
```
# Enable backup_image = yes only in some specific tests, such as
#    unattended_install. In all other tests, it should be no, so keep the global
#    as is.
```
and actually I have not found any occurrence of `backup_image'` or `backup_image"` in projects avocado, avocado-vt, tp-qemu or tp-libvirt - maybe it's actually obsolete and should be removed?

`backup_image_before_test`
-----------------------------------------
Available in config section `vt.setup` section of file `avocado_vt/conf.d/vt.conf`
I can't see the default value. But on one hand, we can assume it's `False` because the sample configuration implies to comment out the `True` option:
```
[vt.setup]
# Backup image before testing (if not already backed up)
#backup_image_before_test = True
# Restore image after testing (if backup present)
#restore_image_after_test = True
# Keep guest running between tests (faster, but unsafe)
#keep_guest_running = False
```
Another indication is that in our CI project files you'll find that it's set explicitly to `True`, although likely not effective (because it's not commented out).

`backup_image_before_testing`
---------------------------------------------
For our CI aarch64 job configurations I see that the `avocado_vt/options.py` is updated to set `backup_image_before_testing` to `no`.

So at this point, I'm really very confused about how your environment got those values. Digging further I see that

`avocado_vt/options.py` per default sets `yes` for this parameter IF `vt.config` section wasn't defined but `vt.setup.backup_image_before_test` is `True`

```
    def _process_restore_image(self):
        if not get_opt(self.config, "vt.config"):
            if get_opt(self.config, "vt.setup.backup_image_before_test"):
                self.cartesian_parser.assign("backup_image_before_testing", "yes")
            if get_opt(self.config, "vt.setup.restore_image_after_test"):
                self.cartesian_parser.assign("restore_image_after_testing", "yes")
```

but also that many tests apparently set this value explicitly.

Conclusion
----------------

You are right, env_process uses `backup_image_before_testing` to determine if to back up (same for the restore variants). So, you can likely leave things as they are for your current setup. Given however the above picture, I'd personally prefer your test confirms that `backup_image_before_testing` really is `yes` because the upgrade is too big of a change for me.

rh-jugraham replied:
Oh, I understand now. I did not realize there were so many params with such similar names - maybe sometime in the future, cleaning up the params would be a helpful task. Since my test does rely on backup_image_before_testing being yes and I don't know exactly where it is being set, I will also set it (and restore_image_after_testing) explicitly.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+def run(test, params, env):
+    """
+    This case verifies that an in place upgrade on a guest succeeds
+    This test only supports upgrading a RHEL X-1 guest to RHEL X on a RHEL X host
```
You might want to add that it also confirms that the page size isn't changed. In a later PR somebody could expand this to 'test upgrade invariants' creating a dictionary before and after the upgrade and confirming they have the same values, if not reporting them.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+    :params target_release: target release version (upgrading to version)
+    """
+    source_major_release = int(source_release.partition(".")[0])
+    target_major_release = int(target_release.partition(".")[0])
```
Is there a reason why you are using `partition` instead of `split`? TBH, I see this function for the first time. Usually, in the framework we use `split`. Out of convention, I'd prefer you used it, too, but I wouldn't insist on it.

rh-jugraham replied:
I honestly don't remember why I chose partition - it may have made more sense in an older version of this code? So split works well.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+    upgrade_repos_path = params.get("upgrade_repos_path")
+    vm_arch_name = params.get("vm_arch_name")
+
+    target_major_release = target_release.partition(".")[0]
```
s. above

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+    """)
+
+    test.log.debug("Temporarily creating leapp_upgrade_repositories file on the host: %s" % upgrade_repos_content)
+    tmp_dir = data_dir.get_tmp_dir()
```
If you use `test.debugdir` instead the file will be stored in the test result folder (e.g. `job-results/latest/test-results/<test-name>/libvirt_upgrade_repositories.repo`) which is the first place where I would check for the file. The `get_tmp_dir` often points to a path below `/var/tmp` which is test-run specific but not as easy to identify as the test run folder.

smitterl commented on libvirt/tests/src/in_place_upgrade_guest.py:
```diff
@@ -0,0 +1,240 @@
+
+    post_page_size = session.cmd_output(pagesize_check_cmd).strip()
+    if target_page_size != post_page_size:
+        # Note: the following is a workaround that should be removed after the leapp tool provides 64k support on aarch64
```
In order for this to be valid for other archs you should add an arch switch:

```
import platform
...

if platform.machine() == "aarch64":
...

```

rh-jugraham replied:
I used vm_arch_name (which is just from params) but good call, it is easy to forget about other architectures

smitterl commented:
Considering that the SUT here is not libvirt but Leapp in KVM guests, you could also consider creating a new test type and moving these files there. E.g. "tp-libvirt/ipu/tests/(cfg|src)/..."
How to do this: https://github.com/autotest/tp-libvirt/pull/3923

Maybe it would be too much effort to just run a single test scenario. I don't know. In that case, maybe the test type could be "misc" and could also host 

https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/security/virt_what_cvm.py

I am not sure about this. Just mentioning it for consideration.

rh-jugraham commented:
> Considering that the SUT here is not libvirt but Leapp in KVM guests, you could also consider creating a new test type and moving these files there. E.g. "tp-libvirt/ipu/tests/(cfg|src)/..." How to do this: #3923
> 
> Maybe it would be too much effort to just run a single test scenario. I don't know. In that case, maybe the test type could be "misc" and could also host
> 
> https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/security/virt_what_cvm.py
> 
> I am not sure about this. Just mentioning it for consideration.

I wasn't sure as, as you mention, it is just a single test and it may be more effort than it is worth. Given that not many test types currently exist, having one that is just for one test feels out of place to me? I do like the idea of adding other tests to this test type (to both justify it better as well as clean up the repo more) but given that doing so 1) is out of the scope of this specific test and 2) may have some unintended side effects with tests being moved around, my initial reaction is to perhaps keep this test as it is but after this is merged, consider moving this test along with others into a new test type.

rh-jugraham commented:
@smitterl Thanks for the thorough review! I added a few questions to the comments that I was uncertain of

smitterl commented:
Hi @juteman thanks for your patience, please check my replies

rh-jugraham commented:
@smitterl All updated!! Thanks again!
